### PR TITLE
Issue 631 run services with supervisor

### PR DESCRIPTION
**Pull request by johansmitsnl**
Use the supervisor as the main process. The start-mailserver is started
from the supervisord and then this process triggers others.

Defined some default variable in the Dockerfile. In order for
supervisored to build the command lines the ENV variable need to be set.
Therefore the defaults are defined.

Some processes are not single processes like postfix and fail2ban and
they have a wrapper. The wrapper takes care of proper shutdown and checking
if the process is running or not. Supervisored will restart the wrapping
script if the process is gone.

Increased some delays between tests because sometimes they where to short
for all containers to be running.

### CODE REVIEW

**Comment by tomav** [comment]
Restarted build.

File: target/fail2ban-wrapper.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# start fail2ban
+/usr/bin/fail2ban-client start
+
+# lets give fail2ban some time to start
+sleep 5
+
+# wait until fail2ban is dead (triggered by trap)
+while kill -0 "`cat /var/run/fail2ban/fail2ban.pid`"; do
+  sleep 5
+done
```
Comment (tomav) [inline target/fail2ban-wrapper.sh]: Missing newline.

File: target/postfix-wrapper.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# start postfix
+service postfix start
+
+# lets give postfix some time to start
+sleep 5
+
+# wait until postfix is dead (triggered by trap)
+while kill -0 "`cat /var/spool/postfix/pid/master.pid`"; do
+  sleep 5
+done
```
Comment (tomav) [inline target/postfix-wrapper.sh]: Missing newline

File: target/start-mailserver.sh [hunk 3]
```diff
@@ -1124,60 +1112,62 @@ function start_daemons() {
...
 	# Here we are starting sasl and imap, not pop3 because it's disabled by default
+
 	notify 'task' 'Starting dovecot services' 'n'
-	display_startup_daemon "/usr/sbin/dovecot -c /etc/dovecot/dovecot.conf"
 
 	if [ "$ENABLE_POP3" = 1 ]; then
 		notify 'task' 'Starting pop3 services' 'n'
 		mv /etc/dovecot/protocols.d/pop3d.protocol.disab /etc/dovecot/protocols.d/pop3d.protocol
-		display_startup_daemon "/usr/sbin/dovecot reload"
+		# /usr/sbin/dovecot reload
```
Comment (tomav) [inline target/start-mailserver.sh]: If useless, this should be removed.

File: target/start-mailserver.sh [hunk 4]
```diff
@@ -1124,60 +1112,62 @@ function start_daemons() {
...
 	fi
 
 	if [ -f /tmp/docker-mailserver/dovecot.cf ]; then
 		cp /tmp/docker-mailserver/dovecot.cf /etc/dovecot/local.conf
-		/usr/sbin/dovecot reload
+		# /usr/sbin/dovecot reload
```
Comment (tomav) [inline target/start-mailserver.sh]: If useless, this should be removed.

File: target/supervisor/saslauth.conf [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+pidfile=/var/run/saslauthd/saslauthd.pid
+
+[program:saslauthd_shadow]
+startsecs=0
+autostart=false
+autorestart=true
+stdout_logfile=/var/log/supervisor/%(program_name)s.log
+stderr_logfile=/var/log/supervisor/%(program_name)s.log
+command=/usr/sbin/saslauthd -d -a shadow -O %(ENV_SASLAUTHD_MECH_OPTIONS)s
+pidfile=/var/run/saslauthd/saslauthd.pid
```
Comment (tomav) [inline target/supervisor/saslauth.conf]: Missing newline

File: target/supervisor/supervisor-app.conf [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+user=fetchmail
+command=/usr/bin/fetchmail -f /etc/fetchmailrc -v --nodetach --daemon 300 -i /var/lib/fetchmail/.fetchmail-UIDL-cache --pidfile /var/run/fetchmail/fetchmail.pid
+
+[program:postfix]
+startsecs=0
+autostart=false
+autorestart=true
+stdout_logfile=/var/log/supervisor/%(program_name)s.log
+stderr_logfile=/var/log/supervisor/%(program_name)s.log
+command=/usr/local/bin/postfix-wrapper.sh
```
Comment (tomav) [inline target/supervisor/supervisor-app.conf]: Missing newline

File: test/tests.bats [hunk 7]
```diff
@@ -1187,3 +1183,53 @@ load 'test_helper/bats-assert/load'
...
+
+@test "checking restart of process: clamav (clamav disabled by ENABLED_CLAMAV=0)" {
+  run docker exec mail_disabled_clamav_spamassassin /bin/bash -c "pkill -f clamd && sleep 10 && ps aux --forest | grep -v grep | grep '/usr/sbin/clamd'"
+  assert_failure
+}
+
+@test "checking restart of process: saslauthd (saslauthd server enabled)" {
+  run docker exec mail_with_ldap /bin/bash -c "pkill saslauthd && sleep 10 && ps aux --forest | grep -v grep | grep '/usr/sbin/saslauthd'"
+  assert_success
+}
```
Comment (tomav) [inline test/tests.bats]: Missing newline

File: Makefile [hunk 8]
```diff
@@ -139,6 +146,7 @@ run:
 		-v "`pwd`/test":/tmp/docker-mailserver-test \
 		-e ENABLE_POSTFIX_VIRTUAL_TRANSPORT=1 \
 		-e POSTFIX_DAGENT=lmtp:127.0.0.1:24 \
+		-e DMS_DEBUG=0 \
 		-h mail.my-domain.com -t $(NAME)
 	sleep 30
```
Comment (tomav) [inline Makefile]: If sleep increased too 40 at the end, perhaps this one can be set to 15 like previous.

File: Makefile [hunk 9]
```diff
@@ -130,6 +136,7 @@ run:
 		-e SASLAUTHD_MECHANISMS=rimap \
 		-e SASLAUTHD_MECH_OPTIONS=127.0.0.1 \
 		-e POSTMASTER_ADDRESS=postmaster@localhost.localdomain \
+		-e DMS_DEBUG=0 \
 		-h mail.my-domain.com -t $(NAME)
 	# Wait for containers to fully start
```
Comment (tomav) [inline Makefile]: Could you remove this comment?

**Comment by johansmitsnl** [comment]
@tomav pushed the changes you requested.

**Comment by johansmitsnl** [comment]
@tomav I set the delay back to 20 seconds but some first check (and now to) fails when it is to low. I think it should be 40 (is was good for my testing, 30 didn't do it for me).

**Comment by johansmitsnl** [comment]
@tomav @alinmear I have updated the PR with tests and they pass all now. 
Can someone review and accept if ok?

**Comment by tomav** [comment]
Thanks!
I'll release a new version soon with these changes.

**Comment by dolanor** [comment]
Hello,

Just to know, every service runs by/with supervisor on the latest image?
So using `service fail2ban stop` is useless, right?

**Comment by johansmitsnl** [comment]
@dolanor correct. You should use the supervisorctl stop <service>

**Pull request merged**